pallets/rewards-treasury/src/lib.rs

```diff
@@ -0,0 +1,99 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+#[cfg(test)]
+pub mod mock;
+#[cfg(any(test, feature = "runtime-benchmarks"))]
+mod stub;
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+#[frame_support::pallet]
+pub mod pallet {
+    use frame_support::{
+        log,
+        pallet_prelude::*,
+        traits::{tokens::fungible::Mutate, Get},
+    };
+    use frame_system::pallet_prelude::BlockNumberFor;
+    use pallet_balances;
+    use sp_std::prelude::*;
+
+    use crate::*;
+
+    /// Configure the pallet by specifying the parameters and types on which it depends.
+    #[pallet::config]
+    pub trait Config<I: 'static = ()>: frame_system::Config + pallet_balances::Config<I> {
+        /// Because this pallet emits events, it depends on the runtime's definition of an event.
+        type RuntimeEvent: From<Event<Self, I>>
+            + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+        /// The epoch length in blocks. At each epoch's end the penultimate (last but one) balance is burnt.
+        type Epoch: Get<<Self as frame_system::Config>::BlockNumber>;
+        /// The ID for this pallet
+        #[pallet::constant]
+        type Treasury: Get<<Self as frame_system::Config>::AccountId>;
+    }
+
+    #[pallet::storage]
+    #[pallet::getter(fn penultimate_balance)]
+    pub(super) type PenultimateBalance<T: Config<I>, I: 'static = ()> =
+        StorageValue<_, T::Balance, ValueQuery>;
+
+    #[pallet::pallet]
+    pub struct Pallet<T, I = ()>(PhantomData<(T, I)>);
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config<I>, I: 'static = ()> {
+        /// A vester started vesting. [amount_burnt]
+        BurntFromTreasuryAtEndOfEpoch(T::Balance),
+    }
+
+    // Errors inform users that something went wrong.
+    #[pallet::error]
+    pub enum Error<T, I = ()> {}
+
+    #[pallet::hooks]
+    impl<T: Config<I>, I: 'static> Hooks<BlockNumberFor<T>> for Pallet<T, I> {
+        fn on_finalize(_: BlockNumberFor<T>) {}
+
+        fn on_initialize(current_block: T::BlockNumber) -> Weight {
+            if current_block % T::Epoch::get() == 0u16.into() {
```

Comment:
I understand the rational behind this simplification, but it will cause unexpected behaviors if we update the `Epoch` type with a runtime upgrade.

For example:

The current `Epoch` value is `10000` and I increase it to `15000`. This will cause the burn to happen twice in 15000 blocks (at 10000 and at 15000), while the expected behavior would be for it to happen at 10000 and at 25000.

```suggestion
            if current_block % T::Epoch::get() == 0u16.into() {
```

Reply:
True, we could introduce another config to specify a block where only after that block a burn will happen. Only that with this the change of config is still a time-sensitive endeavour and a runtime upgrade might wait for approval. to not further increase complexity I would say we just say it's ok if one epoch is shorter/longer.

Reply:
An easy way to solve this is to just add a new storage value with the latest block (initial value = 0) where the latest burn happened and then validate `current_block == <Latest_burn> + T::Epoch::get()`

Reply:
This is also not robust when `T::Epoch` is shortened; it would never be true again. I did combine the ideas so far